android/src/main/java/com/liferay/mobile/android/auth/oauth2/TokenRequester.java

```diff
@@ -0,0 +1,190 @@
+package com.liferay.mobile.android.auth.oauth2;
+
+import com.liferay.mobile.android.auth.SessionCallback;
+import com.liferay.mobile.android.exception.ServerException;
+import com.liferay.mobile.android.http.Headers;
+import com.liferay.mobile.android.service.Session;
+import com.liferay.mobile.android.service.SessionImpl;
+import com.squareup.okhttp.Call;
+import com.squareup.okhttp.Callback;
+import com.squareup.okhttp.Credentials;
+import com.squareup.okhttp.FormEncodingBuilder;
+import com.squareup.okhttp.OkHttpClient;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.RequestBody;
+import com.squareup.okhttp.Response;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class TokenRequester {
+
+	public TokenRequester(Session session, String tokenUrl, String clientId, String clientSecret, List<String> scopes,
+		Map<String, String> parameters) {
+		this.session = session;
+		this.tokenUrl = tokenUrl;
+		this.clientId = clientId;
+		this.clientSecret = clientSecret;
+		this.scopes = scopes;
+		this.parameters = parameters;
+	}
+
+	public static Map<String, String> parametersUsernameAndPassword(final String username, final String password) {
+		return new HashMap<String, String>() {{
+			put("username", username);
+			put("password", password);
+			put(GRANT_TYPE, GRANT_TYPE_PASSWORD);
+		}};
+	}
+
+	public static Map<String, String> parametersClientCredentials() {
+		return new HashMap<String, String>() {{
+			put(GRANT_TYPE, GRANT_TYPE_CLIENT_CREDENTIALS);
+		}};
+	}
+
+	public static Map<String, String> parametersRefresh(final String refreshToken) {
+		return new HashMap<String, String>() {{
+			put("refresh_token", refreshToken);
+			put(GRANT_TYPE, GRANT_TYPE_REFRESH_TOKEN);
+		}};
+	}
+
+	public Session send(final SessionCallback callback) throws Exception {
+		OkHttpClient client = new OkHttpClient();
+		FormEncodingBuilder formBody = new FormEncodingBuilder();
+		Request.Builder requestBuilder = new Request.Builder().url(tokenUrl);
+
+		for (Map.Entry<String, String> parameter : parameters.entrySet()) {
+			formBody.add(parameter.getKey(), parameter.getValue());
+		}
+
+		if (!scopes.isEmpty()) {
+			formBody.add("scope", spaceSeparatedScopes());
+		}
+
+		if (clientSecret != null && !clientSecret.isEmpty()) {
+			requestBuilder.header(Headers.AUTHORIZATION, Credentials.basic(clientId, clientSecret));
+		}
+		else {
+			formBody.add("client_id", clientId);
+		}
+
```

Comment:
What I was referring earlier was that, if something of this fails it'll launch an exception and won't use the callback. Do we want that? To have to deal with the exception+the callback?